pkg/controller/awsprivatelink/awsprivatelink_controller.go

```diff
@@ -854,48 +858,88 @@ func (r *ReconcileAWSPrivateLink) reconcileVPCEndpoint(awsClient *awsClient,
+	// The VPCE that will end up being configured and recorded in the CD status. May already exist, or may be created here.
+	var configuredVPCE *ec2.VpcEndpoint
+	// The first existing VPCE we find that's in a good state.
+	var firstGoodVPCE *ec2.VpcEndpoint
+	// The total number of VPCEs for our filter.
+	vpceCount := 0
 
 	input := &ec2.DescribeVpcEndpointsInput{
 		Filters: []*ec2.Filter{tag},
 	}
 
-	var allVpcEndpoints []*ec2.VpcEndpoint
 	for {
 		resp, err := awsClient.hub.DescribeVpcEndpoints(input)
 		if err != nil {
 			endpointLog.WithError(err).Error("error getting VPC Endpoint")
 			return modified, nil, err
 		}
 
-		allVpcEndpoints = append(allVpcEndpoints, resp.VpcEndpoints...)
+		vpceCount += len(resp.VpcEndpoints)
+		// Only bother searching if we haven't already found the ones we're looking for
+		if firstGoodVPCE == nil || configuredVPCE == nil {
+			for _, vpce := range resp.VpcEndpoints {
+				if firstGoodVPCE == nil && goodVPCEStates.Has(*vpce.State) {
+					firstGoodVPCE = vpce
+				}
+				if configuredVPCE == nil && *vpce.VpcEndpointId == cd.Status.Platform.AWS.PrivateLink.VPCEndpointID {
+					configuredVPCE = vpce
+				}
+			}
+		}
 
 		if resp.NextToken == nil {
 			break
 		}
 		input.NextToken = resp.NextToken
 	}
 
-	if len(allVpcEndpoints) == 0 {
+	// Has the previously configured VPCE gone rotten?
+	if configuredVPCE != nil && !goodVPCEStates.Has(*configuredVPCE.State) {
+		endpointLog.WithField("vpcEndpointId", *configuredVPCE.VpcEndpointId).WithField("state", *configuredVPCE.State).Warn("previously configured hub VPCEndpoint is in a bad state; replacing")
+		configuredVPCE = nil
+	}
+	// If we don't already have a VPCE (either never did, or it went bad), see if we found another good one already
+	if configuredVPCE == nil && firstGoodVPCE != nil {
+		endpointLog.WithField("vpcEndpointId", *firstGoodVPCE.VpcEndpointId).Info("adopting hub VPCEndpoint in good state")
+		configuredVPCE = firstGoodVPCE
 		modified = true
+	}
+	// If we still don't have a usable VPCE, create one
+	if configuredVPCE == nil {
 		var err error
-		vpcEndpoint, err = r.createVPCEndpoint(awsClient.hub, cd, metadata, vpcEndpointService, logger)
+		endpointLog.Info("creating hub VPCEndpoint")
+		configuredVPCE, err = r.createVPCEndpoint(awsClient.hub, cd, metadata, vpcEndpointService, logger)
 		if err != nil {
-			logger.WithError(err).Error("error creating VPC Endpoint for service")
+			logger.WithError(err).Error("error creating hub VPCEndpoint")
 			return modified, nil, err
 		}
-	} else {
-		vpcEndpoint = allVpcEndpoints[0]
+		endpointLog.WithField("vpcEndpointId", *configuredVPCE.VpcEndpointId).Info("created hub VPCEndpoint")
+		modified = true
+		vpceCount++ // Because we just created one
+	}
+
+	// At this point we can log the VPCE count
+	lvl := log.InfoLevel
+	if vpceCount > 1 {
+		lvl = log.WarnLevel
+	}
+	endpointLog.WithField("numVPCEndpoints", vpceCount).Log(lvl, "counted VPCEndpoints for cluster")
```

Comment:
@dustman9000 will this work for your alerting needs?

Reply:
Yep, this will work if we have a known format and a clear error message.